Amend Http2 metrics

zzmao opened the pull request:
Add http2ResponseFlushTime on server side.
Add pendingTasks metric for client and server

cgtz commented on ambry-network/src/main/java/com/github/ambry/network/http2/Http2NetworkClient.java:
```diff
@@ -75,6 +76,9 @@ public Http2NetworkClient(Http2ClientMetrics http2ClientMetrics, Http2ClientConf
         new StreamChannelInitializer());
     this.http2ClientMetrics = http2ClientMetrics;
     correlationIdInFlightToChannelMap = new ConcurrentHashMap<>();
+    Gauge<Long> nettyPendingTasks = () -> Http2Utils.getNumberOfPendingTasks(eventLoopGroup);
```
Instead of exposing the registry, we usually add a method like `registerNettyPendingTasksGauge(Gauge<Long>)` to the metrics class to register the gauge but not require the other class to handle setting the metric name.

Same comment for `AmbryServer`

cgtz commented on ambry-network/src/main/java/com/github/ambry/network/http2/Http2Utils.java:
```diff
@@ -38,4 +41,24 @@ static RequestInfo releaseAndCloseStreamChannel(Channel streamChannel) {
+        logger.debug("EventLoop-{} pending tasks: {}", index, pendingTasks);
+        index++;
+      } else {
+        logger.warn("EventLoop is not SingleThreadEventExecutor");
```
It looks like the logger in this class is still set to `Http2ClientResponseHandler`

cgtz commented on ambry-network/src/main/java/com/github/ambry/network/http2/Http2Utils.java:
```diff
@@ -38,4 +41,24 @@ static RequestInfo releaseAndCloseStreamChannel(Channel streamChannel) {
+  public static long getNumberOfPendingTasks(final EventLoopGroup EventLoopGroup) {
+    int index = 0;
+    long totalPendingTasks = 0;
+    for (final EventExecutor eventExecutor : EventLoopGroup) {
```
EventLoopGroup -> eventLoopGroup

cgtz commented on ambry-server/src/main/java/com/github/ambry/server/AmbryServer.java:
```diff
@@ -309,6 +312,14 @@ public void startup() throws InstantiationException {
                 http2ClientConfig, metrics, nettyMetrics, http2ServerMetrics, serverSecurityService);
         nettyHttp2Server = nioServerFactory.getNioServer();
         nettyHttp2Server.start();
+        Gauge<Long> nettyWorkerPendingTasks =
```
Could this guage registration be done from inside of `StrorageServerNettyFactory`?

jsjtzyy commented on ambry-network/src/main/java/com/github/ambry/network/http2/Http2NetworkClient.java:
```diff
@@ -75,6 +76,9 @@ public Http2NetworkClient(Http2ClientMetrics http2ClientMetrics, Http2ClientConf
         new StreamChannelInitializer());
     this.http2ClientMetrics = http2ClientMetrics;
     correlationIdInFlightToChannelMap = new ConcurrentHashMap<>();
+    Gauge<Long> nettyPendingTasks = () -> Http2Utils.getNumberOfPendingTasks(eventLoopGroup);
+    http2ClientMetrics.registry.register(MetricRegistry.name(Http2NetworkClient.class, "NettyPendingTasks"),
+        nettyPendingTasks);
```
This piece of code can be moved to `Http2ClientMetrics`. You can either initialize `Http2ClientMetrics` within `Http2NetworkClient` (thus you can pass in Http2Client or eventLoopGroup) or  you can create a method something like `initializeEventLoopGroupMetrics(eventLoopGroup)` in `Http2ClientMetrics` and call it here.

jsjtzyy commented on ambry-server/src/main/java/com/github/ambry/server/AmbryServer.java:
```diff
@@ -309,6 +312,14 @@ public void startup() throws InstantiationException {
                 http2ClientConfig, metrics, nettyMetrics, http2ServerMetrics, serverSecurityService);
         nettyHttp2Server = nioServerFactory.getNioServer();
         nettyHttp2Server.start();
+        Gauge<Long> nettyWorkerPendingTasks =
+            () -> Http2Utils.getNumberOfPendingTasks(((NettyServer) nettyHttp2Server).getWorkerEventLoopGroup());
+        registry.register(MetricRegistry.name(Http2ServerMetrics.class, "NettyWorkerPendingTasks"),
+            nettyWorkerPendingTasks);
+        Gauge<Long> nettyBossPendingTasks =
+            () -> Http2Utils.getNumberOfPendingTasks(((NettyServer) nettyHttp2Server).getBossEventLoopGroup());
+        registry.register(MetricRegistry.name(Http2ServerMetrics.class, "NettyBossPendingTasks"),
+            nettyBossPendingTasks);
```
Same here, probably move these to `http2ServerMetrics`

jsjtzyy reviewed (approved):
Lgtm, after comments are addressed. Thanks for adding these metrics.

zzmao commented:
@cgtz @jsjtzyy   Addressed your comments:
1. `getNumberOfPendingTasks` is moved to Utils.java to make the dependency easier. (RestServer can call it directly)
2. Gauge registrations are moved to `Http2ClientMetrics` and `NettyMetrics`
3. Server's boss/worker registrations are moved NettyServer, both frontend and storage node can have these metrics.

jsjtzyy commented on ambry-network/src/main/java/com/github/ambry/network/http2/Http2ClientMetrics.java:
```diff
@@ -105,4 +108,8 @@ public Http2ClientMetrics(MetricRegistry registry) {
+
+  void registerNettyPendingTasksGauge(Gauge<Long> pendingTasksGetter) {
+    registry.register(MetricRegistry.name(Http2NetworkClient.class, "NettyPendingTasks"), pendingTasksGetter);
+  }
```
Can we make the gauge inside this method and just pass in eventLoopGroup?
```
void registerNettyPendingTasksGauge(EventLoopGroup eventLoopGroup) {
    Gauge<Long> pendingTasksGetter =  () -> Utils.getNumberOfPendingTasks(eventLoopGroup);
    registry.register(MetricRegistry.name(Http2NetworkClient.class, "NettyPendingTasks"), pendingTasksGetter);
}
```

jsjtzyy commented on ambry-rest/src/main/java/com/github/ambry/rest/NettyMetrics.java:
```diff
@@ -353,4 +353,20 @@ void registerConnectionsStatsHandler(final AtomicLong openConnectionsCount) {
+   * Registers pendingTasks of NettyServer worker's EventLoopGroup to track
+   * @param pendingTasksGetter the pending tasks to tracked
+   */
+  void registerNettyWorkerPendingTasksGauge(Gauge<Long> pendingTasksGetter) {
+    metricRegistry.register(MetricRegistry.name(NettyServer.class, "NettyWorkerPendingTasks"), pendingTasksGetter);
+  }
+
+  /**
+   * Registers pendingTasks of NettyServer boss's EventLoopGroup to track
+   * @param pendingTasksGetter the pending tasks to tracked
+   */
+  void registerNettyBossPendingTasksGauge(Gauge<Long> pendingTasksGetter) {
+    metricRegistry.register(MetricRegistry.name(NettyServer.class, "NettyBossPendingTasks"), pendingTasksGetter);
```
same here, you probably can combine two into single one by adding a string argument in the method:
```
void registerNettyPendingTasksGauge(EventLoopGroup group, String groupName)
{
...
}
```

jsjtzyy reviewed (approved):
lgtm after minor comments are addressed

jsjtzyy commented on ambry-rest/src/main/java/com/github/ambry/rest/NettyMetrics.java:
```diff
@@ -353,4 +355,15 @@ void registerConnectionsStatsHandler(final AtomicLong openConnectionsCount) {
+   */
+  void registerNettyPendingTasksGauge(EventLoopGroup eventLoopGroup, String groupName) {
+    Gauge<Long> pendingTasksGetter = () -> Utils.getNumberOfPendingTasks(eventLoopGroup);
+    metricRegistry.register(MetricRegistry.name(NettyServer.class, "NettyPendingTasks" + groupName),
```
nit:
```
metricRegistry.register(MetricRegistry.name(NettyServer.class, "Netty" + groupName + "PendingTasks"), pendingTasksGetter);
```
is more readable.

zzmao replied:
thanks! Changed.

jsjtzyy reviewed (approved):
lgtm
